client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/channel/failover/strategies/FailoverStrategy.java

```diff
@@ -0,0 +1,59 @@
+/**
+ *  Copyright 2014-2016 CyberVision, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.kaaproject.kaa.client.channel.failover.strategies;
+
+import org.kaaproject.kaa.client.channel.failover.FailoverDecision;
+import org.kaaproject.kaa.client.channel.failover.FailoverStatus;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Failover strategy is responsible for producing failover decisions based on failover statuses.
+ */
+public interface FailoverStrategy {
```

Comment:
There is no method that notifies FailoverStrategy that SDK recovered from the failure. This makes it impossible to implement more or less production ready scenario: "If there is a failure to connect to bootstrap servers i will retry it for 3 times and then notify user somehow". Basically you know that there was subsequent failures but you don't know if you ever recovered between failures.